tortoise/tortoise.go

```diff
@@ -904,40 +831,112 @@ func (t *turtle) layerCutoff() types.LayerID {
 	return t.last.Sub(t.Hdist)
 }
 
-func isUndecided(config Config, decided map[types.LayerID]types.BlockID, lid, last types.LayerID) bool {
-	if _, exists := decided[lid]; exists {
-		return false
+func (t *turtle) decodeExceptions(base, ballot *ballotInfo, exceptions types.Votes) {
+	from := base.layer
+	diff := map[types.LayerID]map[types.BlockID]sign{}
+	for vote, bids := range map[sign][]types.BlockID{
+		support: exceptions.Support,
+		against: exceptions.Against,
+	} {
+		for _, bid := range bids {
+			block, exist := t.blockRefs[bid]
+			if !exist {
+				ballot.conditions.votesBeforeBase = true
```

Comment:
when the ballot voted for a block the nodes hasn't seen before, why is `votesBeforeBase` set to true? how is this related to the base ballot?

Reply:
in this case block layer can't be determined since block is not in the sliding window, so it is implicitly below base ballot, since base ballot is in the sliding window